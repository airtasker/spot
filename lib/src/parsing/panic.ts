import ts from "typescript";

export function panic(nodeOrMessage: ts.Node | string) {
  if (typeof nodeOrMessage === "string") {
    return new Error(nodeOrMessage);
  }
  let syntaxKind = "unknown";
  for (const [key, value] of Object.entries(ts.SyntaxKind)) {
    if (nodeOrMessage.kind === value) {
      syntaxKind = key;
      break;
    }
  }
  return new Error(syntaxKind + ": " + JSON.stringify(nodeOrMessage, null, 2));
}
