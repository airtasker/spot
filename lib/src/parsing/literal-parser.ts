import ts from "typescript";
import { panic } from "./panic";

export type Literal =
  | ObjectLiteral
  | ArrayLiteral
  | StringLiteral
  | NumericLiteral
  | BooleanLiteral;

/**
 * Extracts a Literal from a TypeScript expression, or fails if the expression isn't a supported literal.
 */
export function extractLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.Node
): Literal {
  if (ts.isObjectLiteralExpression(expression)) {
    return extractObjectLiteral(sourceFile, expression);
  } else if (ts.isArrayLiteralExpression(expression)) {
    return extractArrayLiteral(sourceFile, expression);
  } else if (ts.isStringLiteral(expression)) {
    return extractStringLiteral(sourceFile, expression);
  } else if (ts.isNumericLiteral(expression)) {
    return extractNumericLiteral(sourceFile, expression);
  } else if (
    expression.kind === ts.SyntaxKind.TrueKeyword ||
    expression.kind === ts.SyntaxKind.FalseKeyword
  ) {
    return extractBooleanLiteral(sourceFile, expression);
  } else {
    throw panic(`Expected a literal, found ${expression.getText(sourceFile)}`);
  }
}

export interface ObjectLiteral {
  kind: "object";
  properties: {
    [key: string]: Literal;
  };
}

export function isObjectLiteral(literal?: Literal): literal is ObjectLiteral {
  return literal !== undefined && literal.kind === "object";
}

function extractObjectLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.ObjectLiteralExpression
): ObjectLiteral {
  const objectLiteral: ObjectLiteral = {
    kind: "object",
    properties: {}
  };
  for (const property of expression.properties) {
    if (!ts.isPropertyAssignment(property)) {
      throw panic(
        `Unsupported property syntax: ${property.getText(sourceFile)}`
      );
    }
    if (!ts.isIdentifier(property.name)) {
      throw panic(
        `The following should be a plain identifier: ${property.name.getText(
          sourceFile
        )}`
      );
    }
    objectLiteral.properties[
      property.name.getText(sourceFile)
    ] = extractLiteral(sourceFile, property.initializer);
  }
  return objectLiteral;
}

export interface ArrayLiteral {
  kind: "array";
  elements: Literal[];
}

export function isArrayLiteral(literal?: Literal): literal is ArrayLiteral {
  return literal !== undefined && literal.kind === "array";
}

function extractArrayLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.ArrayLiteralExpression
): ArrayLiteral {
  return {
    kind: "array",
    elements: expression.elements.map(e => extractLiteral(sourceFile, e))
  };
}

export interface StringLiteral {
  kind: "string";
  text: string;
}

export function isStringLiteral(literal?: Literal): literal is StringLiteral {
  return literal !== undefined && literal.kind === "string";
}

function extractStringLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.StringLiteral
): StringLiteral {
  const literal = expression.getText(sourceFile);
  return {
    kind: "string",
    // TODO: Unescape strings.
    text: literal.substr(1, literal.length - 2)
  };
}

export interface NumericLiteral {
  kind: "number";
  text: string;
}

export function isNumericLiteral(literal?: Literal): literal is NumericLiteral {
  return literal !== undefined && literal.kind === "number";
}

function extractNumericLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.NumericLiteral
): NumericLiteral {
  const literal = expression.getText(sourceFile);
  return {
    kind: "number",
    text: literal
  };
}

export interface BooleanLiteral {
  kind: "boolean";
  value: boolean;
}

export function isBooleanLiteral(literal?: Literal): literal is BooleanLiteral {
  return literal !== undefined && literal.kind === "boolean";
}

function extractBooleanLiteral(
  sourceFile: ts.SourceFile,
  expression: ts.Node
): BooleanLiteral {
  return {
    kind: "boolean",
    value: expression.kind === ts.SyntaxKind.TrueKeyword
  };
}
