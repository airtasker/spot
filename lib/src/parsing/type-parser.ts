import * as ts from "typescript";
import {
  arrayType,
  ArrayType,
  BOOLEAN,
  booleanConstant,
  integerConstant,
  NULL,
  NUMBER,
  ObjectType,
  objectType,
  optionalType,
  STRING,
  stringConstant,
  Type,
  unionType,
  VOID
} from "../models";
import { extractLiteral } from "./literal-parser";
import { panic } from "./panic";

/**
 * Extracts a Type from a TypeScript type alias or interface, or fails if the expression isn't a supported type.
 */
export function extractType(sourceFile: ts.SourceFile, type: ts.Node): Type {
  switch (type.kind) {
    case ts.SyntaxKind.VoidKeyword:
      return VOID;
    case ts.SyntaxKind.StringKeyword:
      return STRING;
    case ts.SyntaxKind.NumberKeyword:
      return NUMBER;
    case ts.SyntaxKind.BooleanKeyword:
      return BOOLEAN;
  }
  if (ts.isTypeLiteralNode(type)) {
    return extractObjectType(sourceFile, type);
  } else if (ts.isArrayTypeNode(type)) {
    return extractArrayType(sourceFile, type);
  } else if (ts.isLiteralTypeNode(type)) {
    const literal = extractLiteral(sourceFile, type.literal);
    switch (literal.kind) {
      case "string": {
        return stringConstant(literal.text);
      }
      case "number": {
        if (!literal.text.match(/^-?\d+$/)) {
          throw panic(
            `Expected an integer, got this instead: ${type.getText(sourceFile)}`
          );
        }
        return integerConstant(parseInt(literal.text));
      }
      case "boolean": {
        return booleanConstant(literal.value);
      }
      default:
        throw panic(
          `Unexpected literal in type definition: ${type.getText(sourceFile)}`
        );
    }
  } else if (ts.isToken(type) && type.kind === ts.SyntaxKind.NullKeyword) {
    return NULL;
  } else if (ts.isUnionTypeNode(type)) {
    return extractUnionType(sourceFile, type);
  } else if (ts.isTypeReferenceNode(type) && ts.isIdentifier(type.typeName)) {
    const typeName = type.typeName.getText(sourceFile);
    if (typeName === "Optional") {
      if (!type.typeArguments || type.typeArguments.length !== 1) {
        throw panic(
          `Expected exacty one type parameter for Optional, got this instead: ${type.getText(
            sourceFile
          )}`
        );
      }
      const typeParameter = type.typeArguments[0];
      return optionalType(extractType(sourceFile, typeParameter));
    }

    switch (typeName) {
      case "Int32":
        return {
          kind: "int32"
        };
      case "Int64":
        return {
          kind: "int64"
        };
      case "Float":
        return {
          kind: "float"
        };
      case "Double":
        return {
          kind: "double"
        };
      case "Date":
        return {
          kind: "date"
        };
      case "DateTime":
        return {
          kind: "date-time"
        };
      default:
        return {
          kind: "type-reference",
          typeName
        };
    }
  } else {
    throw panic(
      `Expected a plain type identifier, got this instead: ${type.getText(
        sourceFile
      )}`
    );
  }
}

export function extractObjectType(
  sourceFile: ts.SourceFile,
  declaration: ts.TypeLiteralNode | ts.InterfaceDeclaration
): ObjectType {
  const extendsTypeNames = [];
  if (ts.isInterfaceDeclaration(declaration)) {
    for (const heritageClause of declaration.heritageClauses || []) {
      for (const type of heritageClause.types) {
        if (
          ts.isExpressionWithTypeArguments(type) &&
          ts.isIdentifier(type.expression)
        ) {
          extendsTypeNames.push(type.expression.getText(sourceFile));
        }
      }
    }
  }
  const properties: {
    [key: string]: Type;
  } = {};
  for (const member of declaration.members) {
    if (
      !member.name ||
      !ts.isIdentifier(member.name) ||
      !ts.isPropertySignature(member) ||
      !member.type
    ) {
      throw panic(
        `Expected a named and typed property, got this instead: ${member.getText(
          sourceFile
        )}`
      );
    }
    let type = extractType(sourceFile, member.type);
    if (member.questionToken) {
      type = optionalType(type);
    }
    properties[member.name.getText(sourceFile)] = type;
  }
  return objectType(properties, extendsTypeNames);
}

export function extractArrayType(
  sourceFile: ts.SourceFile,
  declaration: ts.ArrayTypeNode
): ArrayType {
  return arrayType(extractType(sourceFile, declaration.elementType));
}

export function extractUnionType(
  sourceFile: ts.SourceFile,
  declaration: ts.UnionTypeNode
): Type {
  return unionType(...declaration.types.map(t => extractType(sourceFile, t)));
}
