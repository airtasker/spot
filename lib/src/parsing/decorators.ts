import * as ts from "typescript";
import { Type } from "../models";
import { extractLiteral, Literal } from "./literal-parser";
import { panic } from "./panic";
import { extractType } from "./type-parser";

/**
 * Extracts the arguments from a single decorator associated with a TypeScript node.
 *
 * For example with the following code:
 * ```
 * @myDecorator(1, 2, 3)
 * function abc() {}
 * ```
 *
 * Then calling extractSingleDecorator() on the function declaration node will return:
 * {
 *   typeParameters: [],
 *   arguments: [1, 2, 3]
 * }
 *
 * This will fail if the same decorator is attached to the same node multiple times.
 * Use extractMultipleDecorators() for that purpose instead.
 *
 * @returns null when there is no decorator with the given name attached to the node.
 */
export function extractSingleDecorator(
  sourceFile: ts.SourceFile,
  node: ts.Node,
  decoratorName: string
): Decorator | null {
  const decorators = extractMultipleDecorators(sourceFile, node, decoratorName);
  if (decorators.length === 1) {
    return decorators[0];
  } else if (decorators.length > 1) {
    throw panic(
      `Expected a single @${decoratorName} decorator, found ${
        decorators.length
      }`
    );
  } else {
    return null;
  }
}

/**
 * Extracts the arguments from a repeated decorator associated with a TypeScript node.
 *
 * For example with the following code:
 * ```
 * @myDecorator(1, 2, 3)
 * @myDecorator("a", "b", "c")
 * function abc() {}
 * ```
 *
 * Then calling extractMultipleDecorators() on the function declaration node will return:
 * [
 *   {
 *     typeParameters: [],
 *     arguments: [1, 2, 3]
 *   },
 *   {
 *     typeParameters: [],
 *     arguments: ["a", "b", "c"]
 *   }
 * ]
 */
export function extractMultipleDecorators(
  sourceFile: ts.SourceFile,
  node: ts.Node,
  decoratorName: string
): Decorator[] {
  const decorators: Decorator[] = [];
  if (node.decorators) {
    for (const decorator of node.decorators) {
      if (
        ts.isIdentifier(decorator.expression) &&
        decorator.expression.getText(sourceFile) === decoratorName
      ) {
        decorators.push({
          typeParameters: [],
          arguments: []
        });
      } else if (ts.isCallExpression(decorator.expression)) {
        if (
          ts.isIdentifier(decorator.expression.expression) &&
          decorator.expression.expression.getText(sourceFile) === decoratorName
        ) {
          decorators.push({
            typeParameters: [...(decorator.expression.typeArguments || [])].map(
              t => extractType(sourceFile, t)
            ),
            arguments: [...decorator.expression.arguments].map(e =>
              extractLiteral(sourceFile, e)
            )
          });
        }
      }
    }
  }
  return decorators;
}

export interface Decorator {
  typeParameters: Type[];
  arguments: Literal[];
}
