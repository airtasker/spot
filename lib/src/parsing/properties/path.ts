import ts from "typescript";
import { DynamicPathComponent, PathComponent, VOID } from "../../models";
import { extractSingleDecorator } from "../decorators";
import {
  isObjectLiteral,
  isStringLiteral,
  ObjectLiteral
} from "../literal-parser";
import { panic } from "../panic";
import { extractParameterType } from "./parameter-type";

/**
 * Returns the path of an endpoint.
 *
 * For example:
 * ```
 * @endpoint({
 *   method: "GET",
 *   path: "/users/:userId"
 * })
 * getUser(@pathParam() userId: string): User {
 *   ...
 * }
 * ```
 *
 * will return a list of path components with:
 * - a static component ("/users/")
 * - a dynamic component named userId of type string
 */
export function extractPath(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  endpointDescription: ObjectLiteral
): PathComponent[] {
  const pathLiteral = endpointDescription.properties["path"];
  if (!isStringLiteral(pathLiteral)) {
    throw panic(
      `Invalid path in endpoint description: ${methodDeclaration.getText(
        sourceFile
      )}`
    );
  }
  const path = pathLiteral.text;
  const pathComponents: PathComponent[] = [];
  const dynamicPathComponents: { [name: string]: DynamicPathComponent } = {};
  if (path.length > 0) {
    let componentStartPosition = 0;
    do {
      if (path.charAt(componentStartPosition) === ":") {
        // The parameter name extends until a character that isn't a valid name.
        const nextNonNamePositionRelative = path
          .substr(componentStartPosition + 1)
          .search(/[^a-z0-9_]/gi);
        const dynamicPathComponent: DynamicPathComponent = {
          kind: "dynamic",
          name: path.substr(
            componentStartPosition + 1,
            nextNonNamePositionRelative === -1
              ? undefined
              : nextNonNamePositionRelative
          ),
          type: VOID,
          description: ""
        };
        pathComponents.push(dynamicPathComponent);
        dynamicPathComponents[dynamicPathComponent.name] = dynamicPathComponent;
        componentStartPosition =
          nextNonNamePositionRelative === -1
            ? -1
            : componentStartPosition + 1 + nextNonNamePositionRelative;
      } else {
        // The static component extends until the next parameter, which starts with ":".
        const nextColumnPosition = path.indexOf(":", componentStartPosition);
        pathComponents.push({
          kind: "static",
          content: path.substring(
            componentStartPosition,
            nextColumnPosition === -1 ? undefined : nextColumnPosition
          )
        });
        componentStartPosition = nextColumnPosition;
      }
    } while (componentStartPosition !== -1);
  }
  for (const parameter of methodDeclaration.parameters) {
    processPathParameter(
      sourceFile,
      methodDeclaration,
      parameter,
      dynamicPathComponents
    );
  }
  return pathComponents;
}

function processPathParameter(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  parameter: ts.ParameterDeclaration,
  dynamicPathComponents: { [name: string]: DynamicPathComponent }
) {
  const pathParamDecorator = extractSingleDecorator(
    sourceFile,
    parameter,
    "pathParam"
  );
  if (!pathParamDecorator) {
    return;
  }
  const name = parameter.name.getText(sourceFile);
  if (dynamicPathComponents[name]) {
    dynamicPathComponents[name].type = extractParameterType(
      sourceFile,
      parameter
    );
  } else {
    throw panic(
      `Found a path parameter that isn't present in path. Expected one of [${Object.keys(
        dynamicPathComponents
      ).join(", ")}], got this instead: ${name}`
    );
  }
  if (pathParamDecorator.arguments.length > 1) {
    throw panic(
      `Expected exactly one or no arguments for @pathParam(), got ${
        pathParamDecorator.arguments.length
      }`
    );
  }
  if (pathParamDecorator.arguments.length === 1) {
    const pathParamDescription = pathParamDecorator.arguments[0];
    if (!isObjectLiteral(pathParamDescription)) {
      throw panic(
        `@pathParam() expects an object literal, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    const descriptionProperty = pathParamDescription.properties["description"];
    if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
      throw panic(
        `@pathParam() expects a string description, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    dynamicPathComponents[name].description = descriptionProperty.text;
  }
}
