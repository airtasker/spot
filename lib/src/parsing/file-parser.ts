import * as fs from "fs-extra";
import * as path from "path";
import * as ts from "typescript";
import { Api } from "../models";
import { validate } from "../validator";
import { parseApiClass } from "./api-class-parser";
import { extractSingleDecorator } from "./decorators";
import { panic } from "./panic";
import { extractObjectType, extractType } from "./type-parser";

/**
 * Parses a TypeScript source file, as well as any other TypeScript files it imports recursively.
 *
 * This should be smart enough not to get stuck into an infinite loop for circular dependencies.
 */
export async function parsePath(sourcePath: string): Promise<Api> {
  const api: Api = {
    endpoints: {},
    types: {}
  };
  await parseFileRecursively(api, new Set(), sourcePath);
  const errors = validate(api);
  if (errors.length > 0) {
    throw panic(errors.join("\n"));
  }
  return api;
}

async function parseFileRecursively(
  api: Api,
  visitedPaths: Set<string>,
  sourcePath: string
): Promise<void> {
  if (!(await fs.existsSync(sourcePath))) {
    if (await fs.existsSync(sourcePath + ".ts")) {
      sourcePath += ".ts";
    } else {
      throw panic(`No source file found at ${sourcePath}`);
    }
  }
  if (
    visitedPaths.has(sourcePath) ||
    path.resolve(sourcePath).startsWith(__dirname)
  ) {
    return;
  } else {
    visitedPaths.add(sourcePath);
  }
  const fileContent = await fs.readFile(sourcePath, "utf8");
  const sourceFile = ts.createSourceFile(
    path.basename(sourcePath),
    fileContent,
    ts.ScriptTarget.Latest
  );
  for (const statement of sourceFile.statements) {
    if (ts.isImportDeclaration(statement)) {
      if (!ts.isStringLiteral(statement.moduleSpecifier)) {
        throw panic(
          `Unsupported import statement: ${statement.moduleSpecifier.getText(
            sourceFile
          )}`
        );
      }
      const importPath = statement.moduleSpecifier.text;
      if (!importPath.startsWith(".")) {
        // This is not a relative import, we'll ignore it.
        continue;
      }
      await parseFileRecursively(
        api,
        visitedPaths,
        path.join(sourcePath, "..", importPath)
      );
    } else if (ts.isClassDeclaration(statement)) {
      const apiDecorator = extractSingleDecorator(sourceFile, statement, "api");
      if (apiDecorator) {
        parseApiClass(sourceFile, statement, api);
      }
    } else if (ts.isTypeAliasDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractType(sourceFile, statement.type);
    } else if (ts.isInterfaceDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractObjectType(sourceFile, statement);
    }
  }
}
