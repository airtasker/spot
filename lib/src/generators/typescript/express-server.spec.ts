import * as fs from "fs-extra";
import * as path from "path";
import { parsePath } from "../../parsing/file-parser";
import {
  generateEndpointHandlerSource,
  generateExpressServerSource
} from "./express-server";

const EXAMPLES_DIR = path.join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "examples",
  "src"
);

describe("TypeScript Express server generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, `${testCaseName}-api.ts`)
        );
        const serverSource = generateExpressServerSource(api);
        expect(serverSource).toMatchSnapshot("server");
        for (const [endpointName, endpoint] of Object.entries(api.endpoints)) {
          const endpointSource = generateEndpointHandlerSource(
            api,
            endpointName,
            endpoint
          );
          expect(endpointSource).toMatchSnapshot(`endpoint:${endpointName}`);
        }
      });
    }
  });
});
