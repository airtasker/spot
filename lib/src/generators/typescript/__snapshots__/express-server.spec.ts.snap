// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:createUser 1`] = `
"import { CreateUserRequest, CreateUserResponse } from \\"../types\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    error?: void;
    status: number;
    data: CreateUserResponse;
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint createUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:deleteUser 1`] = `
"export async function deleteUser(userId: number, authToken: string): Promise<{
    error?: void;
    status: number;
    data: null;
} | {
    error: \\"forbidden\\";
    status?: void;
    data: {
        message: string;
        signedInAs: string;
    };
} | {
    error: string;
    status?: number;
    data: {
        message: string;
    };
}> {
    throw new Error(\\"Endpoint deleteUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:findUsers 1`] = `
"export async function findUsers(): Promise<{
    error?: void;
    status: number;
    data: {
        name: string;
        age?: number;
    }[];
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint findUsers is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: endpoint:getUser 1`] = `
"export async function getUser(userId: number): Promise<{
    error?: void;
    status: number;
    data: {
        name: string;
        age?: number;
    };
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint getUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code multi-file: server 1`] = `
"import * as cors from \\"cors\\";

import * as express from \\"express\\";

import * as validators from \\"./validators\\";

import { createUser } from \\"./endpoints/createUser\\";

import { deleteUser } from \\"./endpoints/deleteUser\\";

import { findUsers } from \\"./endpoints/findUsers\\";

import { getUser } from \\"./endpoints/getUser\\";

const PORT = 3020;

const app = express();

app.use(cors());

app.use(express.json());

app.post(\\"/users/create\\", async (req, res) => {
    const request = req.body;
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await createUser(request, authToken);
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint createUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.delete(\\"/users/:userId-confirmed\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await deleteUser(userId, authToken);
        if (response.error === \\"forbidden\\") {
            if (!validators.validateDeleteUser_specificErrorForbidden(response.data)) {
                throw new Error(\`Invalid error response for specific error forbidden: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(403);
            res.json(response.data);
            return;
        }
        if (\\"error\\" in response) {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid error response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 500);
            res.json(response.data);
        }
        else {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint deleteUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users\\", async (req, res) => {
    try {
        const response = await findUsers();
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateFindUsers_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint findUsers threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users/:userId\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    try {
        const response = await getUser(userId);
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint getUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.listen(PORT, () => {
    console.log(\`Listening on port \${PORT}\`);
});"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:createUser 1`] = `
"import { CreateUserRequest, CreateUserResponse } from \\"../types\\";

export async function createUser(request: CreateUserRequest, authToken: string | void): Promise<{
    error?: void;
    status: number;
    data: CreateUserResponse;
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint createUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:deleteUser 1`] = `
"export async function deleteUser(userId: number, authToken: string): Promise<{
    error?: void;
    status: number;
    data: null;
} | {
    error: \\"forbidden\\";
    status?: void;
    data: {
        message: string;
        signedInAs: string;
    };
} | {
    error: string;
    status?: number;
    data: {
        message: string;
    };
}> {
    throw new Error(\\"Endpoint deleteUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:findUsers 1`] = `
"export async function findUsers(): Promise<{
    error?: void;
    status: number;
    data: {
        name: string;
        age?: number;
    }[];
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint findUsers is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: endpoint:getUser 1`] = `
"export async function getUser(userId: number): Promise<{
    error?: void;
    status: number;
    data: {
        name: string;
        age?: number;
    };
} | {
    error: string;
    status?: number;
    data?: void;
}> {
    throw new Error(\\"Endpoint getUser is not yet implemented!\\");
}"
`;

exports[`TypeScript Express server generator produces valid code single-file: server 1`] = `
"import * as cors from \\"cors\\";

import * as express from \\"express\\";

import * as validators from \\"./validators\\";

import { createUser } from \\"./endpoints/createUser\\";

import { findUsers } from \\"./endpoints/findUsers\\";

import { getUser } from \\"./endpoints/getUser\\";

import { deleteUser } from \\"./endpoints/deleteUser\\";

const PORT = 3020;

const app = express();

app.use(cors());

app.use(express.json());

app.post(\\"/users/create\\", async (req, res) => {
    const request = req.body;
    if (!validators.validateCreateUser_request(request)) {
        throw new Error(\`Invalid request: \${JSON.stringify(request, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateCreateUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await createUser(request, authToken);
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateCreateUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint createUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users\\", async (req, res) => {
    try {
        const response = await findUsers();
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateFindUsers_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint findUsers threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.get(\\"/users/:userId\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateGetUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    try {
        const response = await getUser(userId);
        if (\\"error\\" in response) {
            res.status(response.status || 500);
            res.end();
        }
        else {
            if (!validators.validateGetUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint getUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.delete(\\"/users/:userId-confirmed\\", async (req, res) => {
    const userId = req.params.userId;
    if (!validators.validateDeleteUser_paramUserId(userId)) {
        throw new Error(\`Invalid path parameter userId: \${JSON.stringify(userId, null, 2)}\`);
    }
    const authToken = req.headers[\\"Authorization\\"];
    if (!validators.validateDeleteUser_headerAuthToken(authToken)) {
        throw new Error(\`Invalid header Authorization: \${JSON.stringify(authToken, null, 2)}\`);
    }
    try {
        const response = await deleteUser(userId, authToken);
        if (response.error === \\"forbidden\\") {
            if (!validators.validateDeleteUser_specificErrorForbidden(response.data)) {
                throw new Error(\`Invalid error response for specific error forbidden: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(403);
            res.json(response.data);
            return;
        }
        if (\\"error\\" in response) {
            if (!validators.validateDeleteUser_genericError(response.data)) {
                throw new Error(\`Invalid error response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 500);
            res.json(response.data);
        }
        else {
            if (!validators.validateDeleteUser_response(response.data)) {
                throw new Error(\`Invalid successful response: \${JSON.stringify(response.data, null, 2)}\`);
            }
            res.status(response.status || 200);
            res.json(response.data);
        }
    }
    catch (e) {
        console.error(\\"Endpoint deleteUser threw an unexpected error:\\\\n\\", e);
        res.status(500);
        res.json(\\"An unknown server error has occurred.\\");
    }
});

app.listen(PORT, () => {
    console.log(\`Listening on port \${PORT}\`);
});"
`;
